hw/ip/otbn/dv/otbnsim/sim/stats.py

```diff
@@ -0,0 +1,302 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+from collections import Counter
+from typing import Dict, List, Optional, Tuple
+
+from elftools.dwarf.descriptions import describe_form_class
+from elftools.dwarf.dwarfinfo import DWARFInfo
+from elftools.elf.elffile import ELFFile
+from elftools.elf.sections import SymbolTableSection
+from tabulate import tabulate
+
+from .insn import JAL, JALR, LOOP, LOOPI
+from .isa import OTBNInsn
+from .state import OTBNState
+from .trace import Trace
+
+
+class ExecutionStats:
+    def __init__(self) -> None:
+        self.stall_count = 0
+        self.insn_count = 0
+        self.instruction_histo = Counter()
+        self.func_calls = [] # type: List[Dict[str, str]]
+        self.loops = [] # type: List[Dict[str, str]]
+
+    def record_stall(self) -> None:
+        self.stall_count += 1
+
+    def record_insn(self, pc: int, insn: OTBNInsn,
+                    state_before_commit: OTBNState) -> None:
+        '''Collect execution statistics'''
+
+        self.insn_count += 1
+
+        # Instruction histogram
+        self.instruction_histo[insn.insn.mnemonic] += 1
+
+        # Function calls
+        # - Direct function calls: jal x1, <offset>
+        # - Indirect function calls: jalr x1, <grs1>, 0
+        if isinstance(insn, JAL) and insn.grd == 1:
+            call_stack = state_before_commit.peek_call_stack()
+            if call_stack:
+                caller_func = call_stack[0]
+            else:
+                caller_func = state_before_commit.start_address
```

Comment:
This is fine, but it might be easier just to allow `caller_func` to be `None` at top-level, which also means we don't have to store `State.start_address`?

Reply:
Unfortunately not:
- We want the function address of the entry point in the trace.
- The functions which get caller_func == start_address are the ones called from the entry point (i.e. already one level down). I'm still thinking about using `None` for the "call" to the entry point/start address itself.

Reply:
Yeah, I was imagining that the statistics wouldn't explicitly give the entry point. No big deal though from my side.

Reply:
"I was imagining that the statistics wouldn't explicitly give the entry point."

I did have it both ways already and can see reasons for both. I left the entry point out of the call graph for now, but it feels like useful information at times ... I'll revisit when I use that code more.